### PR TITLE
cap single allocation

### PR DESCRIPTION
**Pull request by leirocks**
add cap single recycler allocation in test and checked build to collect the data
also save the raw config flags to easier read in crawler

### CODE REVIEW

File: lib/Common/Memory/Recycler.cpp [hunk 1]
```diff
@@ -1320,11 +1320,22 @@ Recycler::LargeAlloc(HeapInfo* heap, size_t size, ObjectInfoBits attributes)
 {
     Assert((attributes & InternalObjectInfoBitMask) == attributes);
 
-    if (size >= HeapConstants::MaxLargeObjectSize)
+#if ENABLE_DEBUG_CONFIG_OPTIONS
+    size_t limit = GetRecyclerFlagsTable().MaxSingleAllocSizeInMB * 1024 * 1024;
+#else
+    size_t limit = CONFIG_FLAG(MaxSingleAllocSizeInMB) * 1024 * 1024;
```
Comment (digitalinfinity) [inline lib/Common/Memory/Recycler.cpp]: When would we not have a Recycler flags table? And how does this work for release builds? The flag is marked as FLAGNR?

Comment (leirocks) [reply lib/Common/Memory/Recycler.cpp]: I don't want to change the default for free build before we have the data.
in free build it is a constant so preventing from corrupted

File: lib/Common/Core/ConfigFlagsTable.h [hunk 2]
```diff
@@ -502,6 +502,12 @@ namespace Js
 
                 bool            flagPresent[FlagCount];
 
+#if defined(ENABLE_DEBUG_CONFIG_OPTIONS) && CONFIG_PARSE_CONFIG_FILE
+                // save the jscript.config for easier to get the raw input while analyzing dump file
+                char16          rawInput[512];
```
Comment (curtisman) [inline lib/Common/Core/ConfigFlagsTable.h]: This is only for flags from a file, so sdd "file" to the name.

File: lib/Common/Exceptions/ReportError.cpp [hunk 3]
```diff
@@ -145,4 +145,16 @@ _NOINLINE void OutOfMemory_unrecoverable_error()
     ReportFatalException(NULL, E_OUTOFMEMORY, Fatal_OutOfMemory, scenario);
 }
 
+_NOINLINE void RecyclerSingleAllocationLimit_fatal_error()
```
Comment (curtisman) [inline lib/Common/Exceptions/ReportError.cpp]: unrecoverable.

**Comment by jackhorton** [comment]
Just wondering, should it be noted somewhere that we only officially support limits of up to 2GB? There may be code elsewhere in the codebase that makes assumptions about max object size that don't align with an allocation limit of larger than 2GB?

**Comment by leirocks** [comment]
we can. but I don't think that's very important. Javascript users don't usually care about how the object is *internally* allocated. even it exceeds the limit the runtime can workaround by split it to multiple small allocation (obviously we are not doing that though), and 2GB is very big limit, usually we can't hit it in normal code. also, this change does not change the limit, just provide a way to collect the data in test and checked build

**Pull request merged**